File: src/test/java/info/skyblond/archivedag/controller/grpc/GroupServiceImplTest.java
package info.skyblond.archivedag.controller.grpc;

import info.skyblond.archivedag.config.EmbeddedRedisConfiguration;
import info.skyblond.archivedag.config.WebMvcConfig;
import info.skyblond.archivedag.model.EntityNotFoundException;
import info.skyblond.archivedag.model.PermissionDeniedException;
import info.skyblond.archivedag.protos.common.Empty;
import info.skyblond.archivedag.protos.group.GroupDetail;
import info.skyblond.archivedag.protos.group.GroupNameList;
import info.skyblond.archivedag.protos.group.GroupRequestMessage;
import info.skyblond.archivedag.service.impl.GroupService;
import info.skyblond.archivedag.service.impl.UserManagementService;
import info.skyblond.archivedag.util.GeneralKt;
import io.grpc.internal.testing.StreamRecorder;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.ActiveProfiles;

import java.util.List;

import static info.skyblond.archivedag.controller.grpc.GrpcTestUtils.checkEmptyResponse;
import static info.skyblond.archivedag.controller.grpc.GrpcTestUtils.safeExecutable;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest(
        classes = {EmbeddedRedisConfiguration.class, WebMvcConfig.class}
)
@ActiveProfiles("test")
class GroupServiceImplTest {
    @Autowired
    GroupServiceImpl groupController;
    @Autowired
    GroupService groupService;
    @Autowired
    UserManagementService userManagementService;

    @BeforeEach
    void setUp() {
        safeExecutable(() -> this.userManagementService.createUser("test_user_admin", "password"));
        safeExecutable(() -> this.userManagementService.createUser("test_user", "password"));
        safeExecutable(() -> this.userManagementService.deleteUser("test_user_404"));
        safeExecutable(() -> this.groupService.deleteGroup("group_test_group"));
        safeExecutable(() -> this.groupService.deleteGroup("group_test_group2"));
        safeExecutable(() -> this.groupService.deleteGroup("group_test_group_404"));
    }


    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testListAdmin() {
        createGroup("group_test_group", "test_user");
        createGroup("group_test_group2", "test_user");
        GroupRequestMessage request = GroupRequestMessage.newBuilder()
                .setGroupName("test")
                .setSize(20)
                .build();
        StreamRecorder<GroupNameList> responseObserver = StreamRecorder.create();
        groupController.listGroupName(request, responseObserver);
        assertNull(responseObserver.getError());
        assertEquals(1, responseObserver.getValues().size());
        GroupNameList result = responseObserver.getValues().get(0);
        assertEquals(2, result.getGroupNameList().size());
        assertEquals(List.of("group_test_group", "group_test_group2"), result.getGroupNameList());
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testListJoinedGroupAdmin() {
        createGroup("group_test_group", "test_user");
        createGroup("group_test_group2", "test_user");
        joinGroup("group_test_group2", "test_user");
        joinGroup("group_test_group", "test_user");
        GroupRequestMessage request = GroupRequestMessage.newBuilder()
                .setUsername("test_user")
                .setSize(20)
                .build();
        StreamRecorder<GroupNameList> responseObserver = StreamRecorder.create();
        groupController.listJoinedGroup(request, responseObserver);
        assertNull(responseObserver.getError());
        assertEquals(1, responseObserver.getValues().size());
        GroupNameList result = responseObserver.getValues().get(0);
        assertEquals(2, result.getGroupNameList().size());
        assertEquals(List.of("group_test_group", "group_test_group2"), result.getGroupNameList());
    }

    @WithMockUser(username = "test_user", roles = "VIEWER")
    @Test
    void testListJoinedGroupUser() {
        createGroup("group_test_group", "test_user");
        createGroup("group_test_group2", "test_user");
        joinGroup("group_test_group2", "test_user");
        joinGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setSize(20)
                    .build();
            StreamRecorder<GroupNameList> responseObserver = StreamRecorder.create();
        };
        groupController.listJoinedGroup(ref.request, ref.responseObserver);
        assertNull(ref.responseObserver.getError());
        assertEquals(1, ref.responseObserver.getValues().size());
        GroupNameList result = ref.responseObserver.getValues().get(0);
        assertEquals(2, result.getGroupNameList().size());
        assertEquals(List.of("group_test_group", "group_test_group2"), result.getGroupNameList());

        ref.request = GroupRequestMessage.newBuilder()
                .setSize(20)
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.listJoinedGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testListOwnedGroupAdmin() {
        createGroup("group_test_group", "test_user");
        createGroup("group_test_group2", "test_user");
        joinGroup("group_test_group2", "test_user");
        joinGroup("group_test_group", "test_user");
        GroupRequestMessage request = GroupRequestMessage.newBuilder()
                .setUsername("test_user")
                .setSize(20)
                .build();
        StreamRecorder<GroupNameList> responseObserver = StreamRecorder.create();
        groupController.listOwnedGroup(request, responseObserver);
        assertNull(responseObserver.getError());
        assertEquals(1, responseObserver.getValues().size());
        GroupNameList result = responseObserver.getValues().get(0);
        assertEquals(2, result.getGroupNameList().size());
        assertEquals(List.of("group_test_group", "group_test_group2"), result.getGroupNameList());
    }

    @WithMockUser(username = "test_user", roles = "VIEWER")
    @Test
    void testListOwnedGroupUser() {
        createGroup("group_test_group", "test_user");
        createGroup("group_test_group2", "test_user");
        joinGroup("group_test_group2", "test_user");
        joinGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setSize(20)
                    .build();
            StreamRecorder<GroupNameList> responseObserver = StreamRecorder.create();
        };
        groupController.listOwnedGroup(ref.request, ref.responseObserver);
        assertNull(ref.responseObserver.getError());
        assertEquals(1, ref.responseObserver.getValues().size());
        GroupNameList result = ref.responseObserver.getValues().get(0);
        assertEquals(2, result.getGroupNameList().size());
        assertEquals(List.of("group_test_group", "group_test_group2"), result.getGroupNameList());

        ref.request = GroupRequestMessage.newBuilder()
                .setSize(20)
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.listOwnedGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testLeaveGroupAdmin() {
        createGroup("group_test_group", "test_user");
        joinGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .setUsername("test_user")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.leaveGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(EntityNotFoundException.class, () ->
                groupController.joinGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user", roles = "UPLOADER")
    @Test
    void testLeaveGroupUser() {
        createGroup("group_test_group", "test_user");
        joinGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.leaveGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        joinGroup("group_test_group", "test_user_404");
        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        groupController.leaveGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        createGroup("group_test_group2", "test_user_404");
        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.leaveGroup(ref.request, ref.responseObserver));
    }

    private void joinGroup(String groupName, String username) {
        safeExecutable(() -> this.groupService.addUserToGroup(groupName, username));
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testJoinGroupAdmin() {
        createGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.joinGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(EntityNotFoundException.class, () ->
                groupController.joinGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user", roles = "UPLOADER")
    @Test
    void testJoinGroupUser() {
        createGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.joinGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        createGroup("group_test_group2", "test_user_admin");
        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.joinGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testTransferGroupAdmin() {
        createGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .setUsername("test_user_admin")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.transferGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(EntityNotFoundException.class, () ->
                groupController.transferGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user", roles = "UPLOADER")
    @Test
    void testTransferGroupUser() {
        createGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .setUsername("test_user_admin")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.transferGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        createGroup("group_test_group2", "test_user_admin");
        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.transferGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user", roles = "VIEWER")
    @Test
    void testQueryGroup() {
        createGroup("group_test_group", "test_user");
        GroupRequestMessage request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .build();
        StreamRecorder<GroupDetail> responseObserver = StreamRecorder.create();
        groupController.queryGroup(request, responseObserver);
        assertNull(responseObserver.getError());
        assertEquals(1, responseObserver.getValues().size());
        GroupDetail result = responseObserver.getValues().get(0);
        assertEquals("group_test_group", result.getGroupName());
        assertEquals("test_user", result.getOwner());
        long d = GeneralKt.getUnixTimestamp() - result.getCreatedTimestamp();
        assertTrue(d <= 1 && d >= 0);
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testDeleteGroupAdmin() {
        createGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.deleteGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(EntityNotFoundException.class, () ->
                groupController.deleteGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user", roles = "UPLOADER")
    @Test
    void testDeleteGroupUser() {
        createGroup("group_test_group", "test_user");
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.deleteGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        createGroup("group_test_group2", "test_user_404");
        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.deleteGroup(ref.request, ref.responseObserver));

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.deleteGroup(ref.request, ref.responseObserver));
    }

    private void createGroup(String groupName, String owner) {
        this.groupService.createGroup(groupName, owner);
    }

    @WithMockUser(username = "test_user_admin", roles = "ADMIN")
    @Test
    void testCreateGroupAdmin() {
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .setUsername("test_user")
                .build();
        ref.responseObserver = StreamRecorder.create();
        groupController.createGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .setUsername("test_user_404") // 404 user
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(EntityNotFoundException.class, () ->
                groupController.createGroup(ref.request, ref.responseObserver));
    }

    @WithMockUser(username = "test_user", roles = "UPLOADER")
    @Test
    void testCreateGroupUser() {
        var ref = new Object() {
            GroupRequestMessage request = GroupRequestMessage.newBuilder()
                    .setGroupName("group_test_group")
                    .build();
            StreamRecorder<Empty> responseObserver = StreamRecorder.create();
        };
        groupController.createGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .setUsername("test_user")
                .build();
        ref.responseObserver = StreamRecorder.create();
        groupController.createGroup(ref.request, ref.responseObserver);
        checkEmptyResponse(ref.responseObserver);

        ref.request = GroupRequestMessage.newBuilder()
                .setGroupName("group_test_group2")
                .setUsername("test_user_404")
                .build();
        ref.responseObserver = StreamRecorder.create();
        assertThrows(PermissionDeniedException.class, () ->
                groupController.createGroup(ref.request, ref.responseObserver));
    }

}

File: backend/src/test/kotlin/info/skyblond/archivedag/arudaz/controller/grpc/GroupServiceImplTest.kt
package info.skyblond.archivedag.arudaz.controller.grpc

import info.skyblond.archivedag.arstue.GroupService
import info.skyblond.archivedag.arstue.UserManagementService
import info.skyblond.archivedag.arudaz.protos.common.Empty
import info.skyblond.archivedag.arudaz.protos.group.*
import info.skyblond.archivedag.commons.EntityNotFoundException
import info.skyblond.archivedag.commons.PermissionDeniedException
import info.skyblond.archivedag.commons.getUnixTimestamp
import info.skyblond.archivedag.safeExecutable
import io.grpc.internal.testing.StreamRecorder
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.context.ActiveProfiles

@SpringBootTest
@ActiveProfiles("test")
internal class GroupServiceImplTest {
    @Autowired
    lateinit var groupController: GroupServiceImpl

    @Autowired
    lateinit var groupService: GroupService

    @Autowired
    lateinit var userManagementService: UserManagementService

    @BeforeEach
    fun setUp() {
        safeExecutable { userManagementService.createUser("test_user_admin", "password") }
        safeExecutable { userManagementService.createUser("test_user", "password") }
        safeExecutable { userManagementService.deleteUser("test_user_404") }
        safeExecutable { groupService.deleteGroup("group_test_group") }
        safeExecutable { groupService.deleteGroup("group_test_group2") }
        safeExecutable { groupService.deleteGroup("group_test_group_404") }
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testListAdmin() {
        createGroup("group_test_group", "test_user")
        createGroup("group_test_group2", "test_user")
        val request: ListGroupNameRequest = ListGroupNameRequest.newBuilder()
            .setKeyword("test")
            .setSize(20)
            .build()
        val responseObserver: StreamRecorder<GroupNameListResponse> = StreamRecorder.create()
        groupController.listGroupName(request, responseObserver)
        Assertions.assertNull(responseObserver.error)
        Assertions.assertEquals(1, responseObserver.values.size)
        val result: GroupNameListResponse = responseObserver.values[0]
        Assertions.assertEquals(2, result.groupNameList.size)
        Assertions.assertEquals(listOf("group_test_group", "group_test_group2"), result.groupNameList)
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testListJoinedGroupAdmin() {
        createGroup("group_test_group", "test_user")
        createGroup("group_test_group2", "test_user")
        joinGroup("group_test_group2", "test_user")
        joinGroup("group_test_group", "test_user")
        val request: ListJoinedGroupRequest = ListJoinedGroupRequest.newBuilder()
            .setUsername("test_user")
            .setSize(20)
            .build()
        val responseObserver: StreamRecorder<GroupNameListResponse> = StreamRecorder.create()
        groupController.listJoinedGroup(request, responseObserver)
        Assertions.assertNull(responseObserver.error)
        Assertions.assertEquals(1, responseObserver.values.size)
        val result: GroupNameListResponse = responseObserver.values[0]
        Assertions.assertEquals(2, result.groupNameList.size)
        Assertions.assertEquals(listOf("group_test_group", "group_test_group2"), result.groupNameList)
    }


    @WithMockUser(username = "test_user", roles = ["VIEWER"])
    @Test
    fun testListJoinedGroupUser() {
        createGroup("group_test_group", "test_user")
        createGroup("group_test_group2", "test_user")
        joinGroup("group_test_group2", "test_user")
        joinGroup("group_test_group", "test_user")
        var request: ListJoinedGroupRequest = ListJoinedGroupRequest.newBuilder()
            .setSize(20)
            .build()
        var responseObserver: StreamRecorder<GroupNameListResponse> = StreamRecorder.create()

        groupController.listJoinedGroup(request, responseObserver)
        Assertions.assertNull(responseObserver.error)
        Assertions.assertEquals(1, responseObserver.values.size)
        val result: GroupNameListResponse = responseObserver.values[0]
        Assertions.assertEquals(2, result.groupNameList.size)
        Assertions.assertEquals(listOf("group_test_group", "group_test_group2"), result.groupNameList)
        request = ListJoinedGroupRequest.newBuilder()
            .setSize(20)
            .setUsername("test_user_404")
            .build()
        StreamRecorder.create<GroupNameListResponse>().also { responseObserver = it }
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.listJoinedGroup(request, responseObserver)
        }
    }


    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testListOwnedGroupAdmin() {
        createGroup("group_test_group", "test_user")
        createGroup("group_test_group2", "test_user")
        joinGroup("group_test_group2", "test_user")
        joinGroup("group_test_group", "test_user")
        val request: ListOwnedGroupRequest = ListOwnedGroupRequest.newBuilder()
            .setUsername("test_user")
            .setSize(20)
            .build()
        val responseObserver: StreamRecorder<GroupNameListResponse> = StreamRecorder.create()
        groupController.listOwnedGroup(request, responseObserver)
        Assertions.assertNull(responseObserver.error)
        Assertions.assertEquals(1, responseObserver.values.size)
        val result: GroupNameListResponse = responseObserver.values[0]
        Assertions.assertEquals(2, result.groupNameList.size)
        Assertions.assertEquals(listOf("group_test_group", "group_test_group2"), result.groupNameList)
    }

    @WithMockUser(username = "test_user", roles = ["VIEWER"])
    @Test
    fun testListOwnedGroupUser() {
        createGroup("group_test_group", "test_user")
        createGroup("group_test_group2", "test_user")
        joinGroup("group_test_group2", "test_user")
        joinGroup("group_test_group", "test_user")
        var request: ListOwnedGroupRequest = ListOwnedGroupRequest.newBuilder()
            .setSize(20)
            .build()
        var responseObserver: StreamRecorder<GroupNameListResponse> = StreamRecorder.create()
        groupController.listOwnedGroup(request, responseObserver)
        Assertions.assertNull(responseObserver.error)
        Assertions.assertEquals(1, responseObserver.values.size)
        val result: GroupNameListResponse = responseObserver.values[0]
        Assertions.assertEquals(2, result.groupNameList.size)
        Assertions.assertEquals(listOf("group_test_group", "group_test_group2"), result.groupNameList)
        request = ListOwnedGroupRequest.newBuilder()
            .setSize(20)
            .setUsername("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.listOwnedGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testLeaveGroupAdmin() {
        createGroup("group_test_group", "test_user")
        joinGroup("group_test_group", "test_user")
        var request: LeaveGroupRequest = LeaveGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setLeavingMember("test_user")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.leaveGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = LeaveGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setLeavingMember("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(EntityNotFoundException::class.java) {
            groupController.leaveGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user", roles = ["UPLOADER"])
    @Test
    fun testLeaveGroupUser() {
        createGroup("group_test_group", "test_user")
        joinGroup("group_test_group", "test_user")
        var request: LeaveGroupRequest = LeaveGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.leaveGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        joinGroup("group_test_group", "test_user_404")
        request = LeaveGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setLeavingMember("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        groupController.leaveGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        createGroup("group_test_group2", "test_user_404")
        request = LeaveGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .setLeavingMember("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.leaveGroup(request, responseObserver)
        }
    }

    private fun joinGroup(groupName: String, username: String) {
        safeExecutable { groupService.addUserToGroup(groupName, username) }
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testJoinGroupAdmin() {
        createGroup("group_test_group", "test_user")
        var request: JoinGroupRequest = JoinGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.joinGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = JoinGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setNewMember("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(EntityNotFoundException::class.java) {
            groupController.joinGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user", roles = ["UPLOADER"])
    @Test
    fun testJoinGroupUser() {
        createGroup("group_test_group", "test_user")
        var request: JoinGroupRequest = JoinGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.joinGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        createGroup("group_test_group2", "test_user_admin")
        request = JoinGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.joinGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testTransferGroupAdmin() {
        createGroup("group_test_group", "test_user")
        var request: TransferGroupRequest = TransferGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setNewOwner("test_user_admin")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.transferGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = TransferGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setNewOwner("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(EntityNotFoundException::class.java) {
            groupController.transferGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user", roles = ["UPLOADER"])
    @Test
    fun testTransferGroupUser() {
        createGroup("group_test_group", "test_user")
        var request: TransferGroupRequest = TransferGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .setNewOwner("test_user_admin")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.transferGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        createGroup("group_test_group2", "test_user_admin")
        request = TransferGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .setNewOwner("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.transferGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user", roles = ["VIEWER"])
    @Test
    fun testQueryGroup() {
        createGroup("group_test_group", "test_user")
        val request: QueryGroupRequest = QueryGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        val responseObserver: StreamRecorder<GroupDetailResponse> = StreamRecorder.create()
        groupController.queryGroup(request, responseObserver)
        Assertions.assertNull(responseObserver.error)
        Assertions.assertEquals(1, responseObserver.values.size)
        val result: GroupDetailResponse = responseObserver.values[0]
        Assertions.assertEquals("group_test_group", result.groupName)
        Assertions.assertEquals("test_user", result.owner)
        val d: Long = getUnixTimestamp() - result.createdTimestamp
        Assertions.assertTrue(d in 0..1)
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testDeleteGroupAdmin() {
        createGroup("group_test_group", "test_user")
        var request: DeleteGroupRequest = DeleteGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.deleteGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = DeleteGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(EntityNotFoundException::class.java) {
            groupController.deleteGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user", roles = ["UPLOADER"])
    @Test
    fun testDeleteGroupUser() {
        createGroup("group_test_group", "test_user")
        var request: DeleteGroupRequest = DeleteGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.deleteGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        createGroup("group_test_group2", "test_user_404")
        request = DeleteGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.deleteGroup(request, responseObserver)
        }
        request = DeleteGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.deleteGroup(request, responseObserver)
        }
    }

    private fun createGroup(groupName: String, owner: String) {
        groupService.createGroup(groupName, owner)
    }

    @WithMockUser(username = "test_user_admin", roles = ["ADMIN"])
    @Test
    fun testCreateGroupAdmin() {
        var request: CreateGroupRequest = CreateGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .setOwner("test_user")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.createGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = CreateGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .setOwner("test_user_404") // 404 user
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(EntityNotFoundException::class.java) {
            groupController.createGroup(request, responseObserver)
        }
    }

    @WithMockUser(username = "test_user", roles = ["UPLOADER"])
    @Test
    fun testCreateGroupUser() {
        var request: CreateGroupRequest = CreateGroupRequest.newBuilder()
            .setGroupName("group_test_group")
            .build()
        var responseObserver = StreamRecorder.create<Empty>()
        groupController.createGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = CreateGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .setOwner("test_user")
            .build()
        responseObserver = StreamRecorder.create()
        groupController.createGroup(request, responseObserver)
        checkEmptyResponse(responseObserver)
        request = CreateGroupRequest.newBuilder()
            .setGroupName("group_test_group2")
            .setOwner("test_user_404")
            .build()
        responseObserver = StreamRecorder.create()
        Assertions.assertThrows(PermissionDeniedException::class.java) {
            groupController.createGroup(request, responseObserver)
        }
    }
}
